src/main/java/hudson/remoting/FileSystemJarCache.java

```diff
@@ -71,8 +71,13 @@ protected URL retrieve(Channel channel, long sum1, long sum2) throws IOException
                 tmp.renameTo(target);
 
                 if (target.exists()) {
+                    Checksum expected = new Checksum(sum1, sum2);
+                    Checksum actual = Checksum.forFile(target);
+                    if (!expected.equals(actual)) {
+                        throw new IOException("Checksum of retrieved jar doesn't match: " + target);
+                    }
                     // even if we fail to rename, we are OK as long as the target actually exists at this point
-                    // this can happen if two FileSystejarCache instances share the same cache dir
+                    // this can happen if two FileSystemJarCache instances share the same cache dir
```

Comment:
Though I would expect the checksum verification to happen on `tmp` _before_ the rename. Otherwise we leave corrupt content behind for the next call. So:

* check for cache file; if present, use it, stop now
* download to `tmp`
* verify checksum
* try to rename to real file
   * if that succeeded, OK, stop
   * if that failed
      * and the real file is not there, die
      * and it is (a race condition with two caches downloading the same file at the same time), verify that checksum (should be correct, but just in case)